File: sudoku/solver/hint.py
"""Top-level hint search, candidate bookkeeping, and a brute-force solver.

``find_hint`` returns the single simplest applicable step. The working candidate
grid it uses is seeded from the user's pencil marks when they have them, falling
back to candidates derived from board values — that way a hint reflects the state
the user is actually looking at. ``solve`` exists so the web layer can offer
"reveal this cell" and so tests can confirm a hint never contradicts the solution.
"""

from __future__ import annotations

from typing import Optional

from ..model import COORDS, DIGITS, Board, can_reach
from .techniques import CandGrid, TECHNIQUES, Hint, Coord


def working_candidates(board: Board) -> CandGrid:
    """Build the candidate grid the engine reasons over.

    For each empty cell: use the user's pencil marks intersected with the legal
    (value-derived) candidates when marks exist; otherwise use the legal candidates
    directly. Intersecting keeps us sound even if the user pencilled an impossible
    digit.
    """
    cg: CandGrid = {}
    for r, c in COORDS:
        if board.value(r, c) is not None:
            continue
        legal = board.candidates(r, c)
        marks = board.cell(r, c).pencil_marks
        cg[(r, c)] = (marks & legal) if marks else legal
    return cg


def find_hint(board: Board, cg: Optional[CandGrid] = None) -> Optional[Hint]:
    """First applicable technique, simplest first. ``None`` if solved or stuck."""
    if not board.is_valid():
        return None
    if cg is None:
        cg = working_candidates(board)
    for technique in TECHNIQUES:
        hint = technique(board, cg)
        if hint is not None:
            return hint
    return None


def apply_to_candidates(board: Board, cg: CandGrid, hint: Hint) -> CandGrid:
    """Apply a hint to a candidate grid in place and return it.

    A placement removes the cell from the grid and clears that digit from peers; an
    elimination drops the listed digits from the named cells.

    Takes ``board`` because a cell's peers are board data once cages are in play,
    not something derivable from coordinates alone (see ``Board.peers``).
    """
    if hint.action == "place":
        (r, c) = hint.cells[0]
        d = hint.digits[0]
        cg.pop((r, c), None)
        for pr, pc in board.peers(r, c):
            if (pr, pc) in cg:
                cg[(pr, pc)].discard(d)
    else:
        for (r, c) in hint.cells:
            if (r, c) in cg:
                cg[(r, c)] -= set(hint.digits)
    return cg


def apply_hint(board: Board, hint: Hint) -> Board:
    """Return a copy of ``board`` with ``hint`` applied (values for placements,
    pencil marks for eliminations)."""
    new = Board.from_dict(board.to_dict())
    if hint.action == "place":
        (r, c) = hint.cells[0]
        new.set_value(r, c, hint.digits[0])
    else:
        for (r, c) in hint.cells:
            new.cell(r, c).pencil_marks -= set(hint.digits)
    return new


def solve_with_techniques(board: Board) -> tuple[Board, list[Hint], bool]:
    """Repeatedly apply ``find_hint`` until solved or stuck.

    Returns ``(final_board, steps, solved)``. Operates on a persistent candidate
    grid so eliminations accumulate. Placements are written back to the board.
    """
    work = Board.from_dict(board.to_dict())
    cg = working_candidates(work)
    steps: list[Hint] = []
    while True:
        hint = find_hint(work, cg)
        if hint is None:
            break
        steps.append(hint)
        if hint.action == "place":
            (r, c) = hint.cells[0]
            work.set_value(r, c, hint.digits[0])
        else:
            # Mirror the elimination onto the board's own pencil marks, not just
            # the candidate grid. An impossible mark is already absent from `cg`,
            # so without this the same hint would be re-found forever.
            for (r, c) in hint.cells:
                work.cell(r, c).pencil_marks -= set(hint.digits)
        apply_to_candidates(work, cg, hint)
    return work, steps, work.is_solved()


def solve(board: Board) -> Optional[Board]:
    """Backtracking solver. Returns a solved copy, or ``None`` if unsolvable."""
    work = Board.from_dict(board.to_dict())
    if not work.is_valid():
        return None
    if _backtrack(work):
        return work
    return None


def _backtrack(board: Board) -> bool:
    """Depth-first search with cage-sum propagation.

    Two things made the straightforward version unusable on a real Killer board.
    It rebuilt every empty cell's candidate set from the board at every node —
    34 million set constructions, 800 million cell lookups, on a board that took
    108 seconds to crack. And it pruned only by asking whether the cage it had
    just written into was still *arithmetically* completable, which is far weaker
    than asking which digits its remaining cells can still hold: a cage whose sum
    had become unreachable for a particular cell went unnoticed until the search
    wandered all the way into it.

    So candidates are carried incrementally and undone on backtrack, and each
    placement re-prunes its cage down to digits that can still complete the
    remaining total. Killer boards start nearly empty and lean on cage sums for
    most of their constraint, so that propagation is what makes the search finite
    in practice rather than merely in principle.
    """
    cands: dict[Coord, set[int]] = {}
    for r, c in COORDS:
        if board.value(r, c) is None:
            cands[(r, c)] = board.candidates(r, c)
            if not cands[(r, c)]:
                return False
    peers = {cell: tuple(board.peers(*cell)) for cell in cands}

    # Per-cage bookkeeping, kept in step with the board: what the cage still owes,
    # which of its cells are still empty, and which digits it has not used up.
    cage_of: dict[Coord, int] = {}
    owed: list[int] = []
    empty: list[set[Coord]] = []
    unused: list[set[int]] = []
    for i, cage in enumerate(board.cages):
        used, blank = set(), set()
        for cell in cage.cells:
            cage_of[cell] = i
            v = board.value(*cell)
            blank.add(cell) if v is None else used.add(v)
        owed.append(cage.sum - sum(used))
        empty.append(blank)
        unused.append(set(DIGITS) - used)

    def strip(cell: Coord, gone: set[int], undo: list) -> bool:
        """Remove ``gone`` from ``cell``'s candidates; False if nothing survives."""
        hit = cands[cell] & gone
        if not hit:
            return True
        cands[cell] -= hit
        undo.append((cell, hit))
        return bool(cands[cell])

    def prune(i: int, undo: list) -> bool:
        """Cut cage ``i``'s empty cells to digits that can still complete its sum."""
        cells, rem, pool = empty[i], owed[i], frozenset(unused[i])
        k = len(cells)
        if k == 0:
            return rem == 0
        for cell in cells:
            keep = {
                d
                for d in cands[cell]
                if d in pool and can_reach(k - 1, rem - d, pool - {d})
            }
            if not strip(cell, cands[cell] - keep, undo):
                return False
        return True

    opening: list = []
    for i in range(len(board.cages)):
        if not prune(i, opening):
            return False

    def step() -> bool:
        if not cands:
            return True
        cell = min(cands, key=lambda x: len(cands[x]))
        i = cage_of.get(cell)
        for d in sorted(cands[cell]):
            undo: list = []
            saved = cands.pop(cell)
            ok = all(strip(p, {d}, undo) for p in peers[cell] if p in cands)
            touched = ok and i is not None
            if touched:
                empty[i].discard(cell)
                owed[i] -= d
                unused[i].discard(d)
                ok = prune(i, undo)
            if ok:
                board.set_value(*cell, d)
                if step():
                    return True
                board.set_value(*cell, None)
            if touched:
                empty[i].add(cell)
                owed[i] += d
                unused[i].add(d)
            for other, hit in undo:
                cands[other] |= hit
            cands[cell] = saved
        return False

    return step()

File: web/sudoku/hint.ts
// Top-level hint search and candidate bookkeeping. (The solver lives in
// solver.ts.)
//
// `findHint` returns the single simplest applicable step. The working Candidate
// grid it uses is seeded from the player's Pencil marks when they have them,
// falling back to Candidates derived from board values — that way a hint
// reflects the state the player is actually looking at.

import { Board, cellName, idx, rc, type Coord } from "./model.ts";
import { TECHNIQUES, type CandGrid, type Hint } from "./techniques.ts";

/** A hint in the shape the page uses, where a Cell is `{r, c}` rather than a
 * pair. The catalogue itself never sees this — it is the page's shape, and it
 * lives here with the rest of what the page calls. */
export type WireHint = Omit<Hint, "cells"> & { cells: { r: number; c: number }[] };

export function hintToWire(hint: Hint): WireHint {
  return { ...hint, cells: hint.cells.map(([r, c]) => ({ r, c })) };
}

/**
 * Candidates per empty cell, row-major.
 *
 * Row-major because techniques return "the first cell that qualifies", so the
 * insertion order is what decides which cell a hint names — the same order
 * Python's dict of coordinates has.
 */
function candidateGrid(board: Board, seed: (r: number, c: number) => Set<number>): CandGrid {
  const cg: CandGrid = new Map();
  for (let i = 0; i < 81; i++) {
    const [r, c] = rc(i);
    if (board.value(r, c) === null) cg.set(i, seed(r, c));
  }
  return cg;
}

/**
 * The Candidate grid the engine reasons over when it hints.
 *
 * For each empty cell: the player's Pencil marks intersected with the legal
 * (value-derived) Candidates when marks exist; otherwise the legal Candidates
 * directly. Intersecting keeps us sound even if the player pencilled an
 * impossible digit.
 */
export function workingCandidates(board: Board): CandGrid {
  return candidateGrid(board, (r, c) => {
    const legal = board.candidates(r, c);
    const marks = board.cell(r, c).pencilMarks;
    if (marks.size === 0) return legal;
    return new Set([...marks].filter((d) => legal.has(d)));
  });
}

/**
 * The Candidate grid to solve from: derived from values alone.
 *
 * Deliberately blind to Pencil marks. A hint must reason from what the player
 * sees, but a solve must not: a *missing* mark would narrow the search past the
 * real solution.
 */
export function derivedCandidates(board: Board): CandGrid {
  return candidateGrid(board, (r, c) => board.candidates(r, c));
}

/** First applicable technique, simplest first. `null` if solved or stuck. */
export function findHint(board: Board, cg?: CandGrid): Hint | null {
  if (!board.isValid()) return null;
  const grid = cg ?? workingCandidates(board);
  for (const technique of TECHNIQUES) {
    const hint = technique(board, grid);
    if (hint !== null) return hint;
  }
  return null;
}

/**
 * Apply a hint to a Candidate grid in place and return it.
 *
 * A placement removes the cell from the grid and clears that digit from peers;
 * an elimination drops the listed digits from the named cells.
 *
 * Takes `board` because a cell's Peers are board data once Cages are in play,
 * not something derivable from coordinates alone (see `Board.peers`).
 */
export function applyToCandidates(board: Board, cg: CandGrid, hint: Hint): CandGrid {
  if (hint.action === "place") {
    const [r, c] = hint.cells[0];
    const d = hint.digits[0];
    cg.delete(idx(r, c));
    for (const p of board.peers(r, c)) cg.get(p)?.delete(d);
  } else {
    for (const [r, c] of hint.cells) {
      const cell = cg.get(idx(r, c));
      if (cell) for (const d of hint.digits) cell.delete(d);
    }
  }
  return cg;
}

/** Rub a hint's digits out of the Pencil marks it names, in place. */
function eraseMarks(board: Board, hint: Hint): void {
  for (const [r, c] of hint.cells) {
    for (const d of hint.digits) board.cell(r, c).pencilMarks.delete(d);
  }
}

/** A copy of `board` with `hint` applied: values for placements, Pencil marks
 * for eliminations. */
export function applyHint(board: Board, hint: Hint): Board {
  const next = Board.fromWire(board.toWire());
  if (hint.action === "place") {
    const [r, c] = hint.cells[0];
    next.setValue(r, c, hint.digits[0]);
  } else {
    eraseMarks(next, hint);
  }
  return next;
}

export interface TechniqueRun {
  board: Board;
  steps: Hint[];
  solved: boolean;
}

/**
 * Repeatedly apply `findHint` until solved or stuck.
 *
 * Operates on a persistent Candidate grid so eliminations accumulate.
 * Placements are written back to the board.
 */
export function solveWithTechniques(board: Board, cg?: CandGrid): TechniqueRun {
  const work = Board.fromWire(board.toWire());
  const grid = cg ?? workingCandidates(work);
  const steps: Hint[] = [];
  for (;;) {
    const hint = findHint(work, grid);
    if (hint === null) break;
    steps.push(hint);
    if (hint.action === "place") {
      const [r, c] = hint.cells[0] as Coord;
      work.setValue(r, c, hint.digits[0]);
    } else {
      // Mirror the elimination onto the board's own Pencil marks, not just the
      // Candidate grid. An impossible mark is already absent from `grid`, so
      // without this the same hint would be re-found forever.
      eraseMarks(work, hint);
    }
    applyToCandidates(work, grid, hint);
  }
  return { board: work, steps, solved: work.isSolved() };
}

/** The gentlest reveal: which Unit or Cell to look at, without saying what to do.
 *
 * Presentation rather than deduction, which is why it sits here in the Sudoku
 * engine's own module and not in something shared — Queens keeps a separate
 * implementation of the same idea, per ADR 0001. */
export function nudge(hint: Hint): string {
  if (hint.units.length > 0) return `Look at ${hint.units[0]}.`;
  const [r, c] = hint.cells[0];
  return `Look at ${cellName(r, c)}.`;
}
